### ISSUE TITLE
Export from Grayjay to NewPipe

### ISSUE THREAD

**Issue by tdadvocate** [futo-org/grayjay-android#1157]
### Use Cases

This is useful for having multiple versions of backups in case Grayjay were to ever be taken down or shut down for any reason. This makes sure that our subscriptions, groups, watch history, etc is protected no matter what happens.

**Comment by Rorschach1010** [comment]
I am new to grayjay and so far I like it. I was very surprised to see that I can import my newpipe subscriptions. Since I don't use a youtube account, all my subscriptions are in my newpipe app. However, the first thing I check, when I test alternative apps is always how easy it is to return so I don't land in a vendor lock. So I searched for a way to export subscriptions to newpipe but that doesn't seem to be possible. That's very strange if you can import from that same source so imo this feature is long overdue simply because a new user coming from newpipe might not always stick around. Personally I like what I see but not being able to "get out" is kinda a showstopper.
Seems like this was asked way before:
https://github.com/futo-org/grayjay-android/issues/16

So yeah, would be great to get this implemented to give the users security.

**Comment by tdadvocate** [comment]
> I am new to grayjay and so far I like it. I was very surprised to see that I can import my newpipe subscriptions. Since I don't use a youtube account, all my subscriptions are in my newpipe app. However, the first thing I check, when I test alternative apps is always how easy it is to return so I don't land in a vendor lock. So I searched for a way to export subscriptions to newpipe but that doesn't seem to be possible. That's very strange if you can import from that same source so imo this feature is long overdue simply because a new user coming from newpipe might not always stick around. Personally I like what I see but not being able to "get out" is kinda a showstopper.
> Seems like this was asked way before:
> https://github.com/futo-org/grayjay-android/issues/16
> 
> So yeah, would be great to get this implemented to give the users security.

Yeah. I really wish I would have noticed sooner. Because now I have to half restart on New Pipe because I'm tired of the way Gray Jay works. It's a cool app but it's one step away from being anti-consumer which really makes me unable to recommend it to anyone now. And the fact they don't seem to care and try to avoid adding this just shows how much they don't care. That's on top of it being closed Source when touting itself as open source. It's really just gross behaviour that I can't stand behind. Which is really unfortunate as I do truly like a lot of what gray Jay has to offer.

**Comment by Rorschach1010** [comment]
Yeah,
I didn't make the full transition on purpose and I might probably return to newpipe simply because of this single issue. I once made this mistake with my browser bookmarks. After a wee I realized that the app wasn't what I want. Only then I realized that I can't export them. I had to do it all manually over the course of 2 days. Was very ugly.
Since then, I always check for exports (and imports ofc), whenever I try out alternatives. Not being easily able to return is a no-go.

**Comment by Zvonimir-FUTO** [comment]
I would like to clarify that NewPipe does not currently support exporting in a format compatible with GrayJay. However, we have developed a feature within GrayJay to parse and import content from Newpipe effectively. While it is technically feasible for Newpipe to adopt similar functionality, this responsibility should not fall upon our team.

**Comment by tdadvocate** [comment]
Could the parsing not relatively easily be reversed? It feels like a situation where neither team feels like it is their responsibility. and will never be created because of what is essentially a pissing match between teams. I get it if the system you already created for importing can't easily be reversed, but at the same time it does feel a little weird that it wasn't really mentioned at import. I feel like a warning at least would be nice for people who aren't fully committed so that they can make sure not to get ahead of themselves while testing GrayJay. But maybe I think too much about the user experience rather than anything else.

**Comment by ghost** [comment]
@tdadvocate You could try this python script I have created, it will convert GrayJay subscriptions to NewPipe subscriptions. Simply create `script.py` in the directory that includes the `Subscriptions` file (from the exported GrayJay archieve), then open powershell/terminal in that directory and execute the script `python ./script.py`.
```python
import json
import os
import argparse
def main():
    parser = argparse.ArgumentParser("Converts GrayJay subscriptions to newPipe subscriptions in the current directory")
    parser.add_argument("-f","--filename", help=f"name of the subscriptions file in the directory, default: 'Subscriptions'")
    args = parser.parse_args()

    _dir = os.getcwd()
    _fileName = args.filename if args.filename else "Subscriptions"

    try:
        with open(os.path.join(_dir, _fileName), "r") as file:
            subs = json.loads(file.read())
    except FileNotFoundError:
        print(f"'{_fileName}' was not found in {_dir}")
        return 
    except Exception as s:
        print(f"An error has occured: {s}")

    class Subscription():
        def __init__(self, url):
            self.name = ""
            self.service_id = 0
            self.url = url

    ConvertedSubs = []
    for sub in subs:
        ConvertedSubs.append(Subscription(sub))


    class SubscriptionHolder():
        def __init__(self, subs):
            self.subscriptions = subs

    subHolder = SubscriptionHolder(ConvertedSubs)
    jsonData = json.dumps(subHolder, default=lambda o: o.__dict__, indent=1)

    with open(os.path.join(_dir, "converted-subscriptions.json"), "w") as file:
        file.write(jsonData)

if __name__ == "__main__":
    main()
```
